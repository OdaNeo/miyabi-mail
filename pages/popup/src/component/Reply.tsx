import { splitByPrompts, PROMPT } from '@src/utils/tts';
import { useState } from 'react';

export function Reply({ replyOrigin }: { replyOrigin: string }) {
  const { origin, reply, result } = splitByPrompts(replyOrigin);
  const [showOrigin] = useState(true);

  return (
    <>
      <div className="mt-4 border p-2 rounded text-sm whitespace-pre-wrap">
        <div>{PROMPT.ORIGIN}</div>
        <div>{origin}</div>
      </div>
      <div className="mt-4 border p-2 rounded text-sm whitespace-pre-wrap">
        <div>{PROMPT.REPLY}</div>
        <div>{reply}</div>
      </div>
      <div className="mt-4 border p-2 rounded text-sm whitespace-pre-wrap">
        <div>{PROMPT.RESULT}</div>
        <div>{result}</div>
      </div>
      {showOrigin && (
        <div className="mt-4 border p-2 rounded text-sm whitespace-pre-wrap">
          <div>AI返回原文</div>
          <div>{replyOrigin}</div>
        </div>
      )}
    </>
  );
}
