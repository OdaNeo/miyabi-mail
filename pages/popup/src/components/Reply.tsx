import type { PROMPT_KEYS } from '@src/utils/tts';
import { AnimatePresence, motion } from 'framer-motion';
import { Label } from './ui/label';
import { ChevronDown, ChevronUp, ClipboardCopy } from 'lucide-react';
import { Button } from './ui/button';
import { cn } from '@src/lib/utils';
import { useState } from 'react';
import { MessageBox } from './ui/MessageBox';
import { replyStorage, translationStorage } from '@extension/storage';
import { useStorage } from '@extension/shared';
import { useGeneratedStore } from '@src/store/generatedStore';
import { useExpandedSectionStore } from '@src/store/expandedSectionStore';

export function Reply() {
  const reply = useStorage(replyStorage);
  const translation = useStorage(translationStorage);

  const { isGenerated } = useGeneratedStore();
  const { expandedSection, setExpandedSection } = useExpandedSectionStore();

  const [copied, setCopied] = useState(false);

  const toggleSection = (section: PROMPT_KEYS) => {
    setExpandedSection(expandedSection === section ? null : section);
  };

  const handleCopy = () => {
    navigator.clipboard
      .writeText(reply)
      .then(() => {
        setCopied(true);
        setTimeout(() => setCopied(false), 2000);
      })
      .catch(err => {
        console.error('Failed to copy text: ', err);
      });
  };

  return (
    <AnimatePresence>
      {isGenerated && (
        <motion.div
          initial={{ opacity: 0, height: 0 }}
          animate={{ opacity: 1, height: 'auto' }}
          exit={{ opacity: 0, height: 0 }}
          className="space-y-2 overflow-hidden"
        >
          {(['TRANSLATION', 'REPLY'] as PROMPT_KEYS[]).map(section => (
            <div key={section} className="space-y-1">
              <div
                className="flex justify-between items-center cursor-pointer p-1 rounded hover:bg-slate-100 dark:hover:bg-slate-800 transition-colors"
                onClick={() => toggleSection(section)}
                onKeyUp={e => {
                  if (e.key === 'Enter' || e.key === ' ') {
                    toggleSection(section);
                  }
                }}
                role="button"
                tabIndex={0}
              >
                <Label className="text-xs font-medium text-slate-700 dark:text-slate-300">
                  {section === 'TRANSLATION' ? '原文の翻訳：' : '日本語の返信：'}
                </Label>
                <div className="flex items-center space-x-1">
                  {section === 'REPLY' && (
                    <Button
                      variant="outline"
                      size="sm"
                      onClick={e => {
                        e.stopPropagation();
                        handleCopy();
                      }}
                      className={cn(
                        'h-6 text-xs transition-colors',
                        copied
                          ? 'bg-emerald-500 text-white hover:bg-emerald-600'
                          : 'text-slate-600 dark:text-slate-400',
                      )}
                    >
                      {copied ? 'コピー済み' : 'コピー'}
                      <ClipboardCopy className="h-3 w-3 ml-1" />
                    </Button>
                  )}
                  {expandedSection === section ? (
                    <ChevronUp className="h-3 w-3 text-slate-500 dark:text-slate-400" />
                  ) : (
                    <ChevronDown className="h-3 w-3 text-slate-500 dark:text-slate-400" />
                  )}
                </div>
              </div>
              {expandedSection === section && (
                <MessageBox>{section === 'TRANSLATION' ? translation : reply}</MessageBox>
              )}
            </div>
          ))}
        </motion.div>
      )}
    </AnimatePresence>
  );
}
