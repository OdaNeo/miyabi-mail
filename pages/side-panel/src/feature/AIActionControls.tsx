import { Button, Progress, Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from '@extension/ui';
import { Languages, Wand2Icon } from 'lucide-react';
import { useI18n } from '../hooks/useI18n';
import { rolePrompt } from '../utils/tts';
import { useProgress } from '../hooks/useProgress';
import { useGeneratedStore } from '../store/generatedStore';
import { useExpandedSectionStore } from '../store/expandedSectionStore';
import {
  apiVersionStorage,
  languageStorage,
  replyStorage,
  temperatureStorage,
  translationStorage,
} from '@extension/storage';
import { useReplyLoading } from '@/store/replyLoadingStore';
import { useTranslationLoading } from '@/store/translationLoadingStore';
import { usePolishingLoading } from '@/store/polishingLoadingStore';
import { useStorage } from '@extension/shared';
import { LANGUAGES } from '@/const/language';
import { addTaskRecord } from '@/db/taskRecordsRepository';
import { TaskHistoryMapper } from '@/db/Dto';

export function AIActionControls({
  inputText,
  subject,
  setInputText,
  runOpenAIAction,
  handleOnClick,
}: {
  inputText: string;
  subject: string;
  setInputText: (text: string) => void;
  runOpenAIAction: ({
    prompt,
    onStart,
    onFinish,
    onSuccess,
  }: {
    prompt: string;
    onStart?: () => void;
    onFinish?: () => void;
    onSuccess?: (content: string) => void;
  }) => Promise<void>;
  handleOnClick: () => void;
}) {
  const { IS_POLISHING, POLISH, IS_TRANSLATING, TRANSLATION, IS_GENERATING, GENERATE } = useI18n();

  const language = useStorage(languageStorage);

  const { setIsGenerated } = useGeneratedStore();
  const { setExpandedSection } = useExpandedSectionStore();
  const { isReplyLoading, setIsReplyLoading } = useReplyLoading();
  const { isTranslationLoading, setIsTranslationLoading } = useTranslationLoading();
  const { isPolishingLoading, setIsPolishingLoading } = usePolishingLoading();

  const [replyProgress, setReplyProgress] = useProgress(isReplyLoading);
  const [translationProgress, setTranslationProgress] = useProgress(isTranslationLoading);
  const [polishingProgress, setPolishingProgress] = useProgress(isPolishingLoading);

  const isLoading = isReplyLoading || isTranslationLoading || isPolishingLoading;

  const apiVersion = useStorage(apiVersionStorage);
  const temperature = useStorage(temperatureStorage);

  const handleReply = () => {
    const createTime = new Date().toISOString();
    runOpenAIAction({
      prompt: rolePrompt(inputText?.trim(), 'REPLY', language, subject?.trim()),
      onStart: () => {
        setIsGenerated(false), setIsReplyLoading(true), setReplyProgress(0);
      },
      onFinish: () => {
        setIsReplyLoading(false), setReplyProgress(0);
      },
      onSuccess: content => {
        setIsGenerated(true), replyStorage.set(content), setExpandedSection('REPLY');
        addTaskRecord(
          TaskHistoryMapper.toEntity({
            inputText,
            language,
            promptKey: 'REPLY',
            apiVersion,
            temperature,
            subject,
            result: content,
            createTime,
            completedTime: new Date().toISOString(),
          }),
        );
      },
    });
  };

  const handleTranslation = () => {
    const createTime = new Date().toISOString();
    runOpenAIAction({
      prompt: rolePrompt(inputText?.trim(), 'TRANSLATION', language),
      onStart: () => {
        setIsGenerated(false), setIsTranslationLoading(true), setTranslationProgress(0);
      },
      onFinish: () => {
        setIsTranslationLoading(false), setTranslationProgress(0);
      },
      onSuccess: content => {
        setIsGenerated(true), translationStorage.set(content), setExpandedSection('TRANSLATION');
        addTaskRecord(
          TaskHistoryMapper.toEntity({
            inputText,
            language,
            promptKey: 'TRANSLATION',
            apiVersion,
            temperature,
            result: content,
            createTime,
            completedTime: new Date().toISOString(),
          }),
        );
      },
    });
  };

  const handlePolishing = () => {
    const createTime = new Date().toISOString();
    runOpenAIAction({
      prompt: rolePrompt(inputText?.trim(), 'POLISHING'),
      onStart: () => {
        setIsGenerated(false), setIsPolishingLoading(true), setPolishingProgress(0);
      },
      onFinish: () => {
        setIsPolishingLoading(false), setPolishingProgress(0);
      },
      onSuccess: content => {
        setIsGenerated(false), setInputText(content), setExpandedSection(null);
        addTaskRecord(
          TaskHistoryMapper.toEntity({
            inputText,
            language,
            promptKey: 'POLISHING',
            apiVersion,
            temperature,
            result: content,
            createTime,
            completedTime: new Date().toISOString(),
          }),
        );
      },
    });
  };

  return (
    <div
      className="flex items-center px-5 py-3 border-t border-[#e5e7eb] dark:border-[#2e3238] bg-[#f3f4f6] dark:bg-[#23252b]"
      onClick={handleOnClick}
    >
      <Select disabled={isLoading} value={language} onValueChange={languageStorage.set}>
        <SelectTrigger className="w-26 h-9 rounded-md border-[#e5e7eb] dark:border-[#2e3238] bg-white dark:bg-[#1f2128]">
          <SelectValue placeholder="选择语言" />
        </SelectTrigger>
        <SelectContent>
          {LANGUAGES.map(lang => (
            <SelectItem key={lang.value} value={lang.value} className="text-sm">
              {lang.label}
            </SelectItem>
          ))}
        </SelectContent>
      </Select>
      <div className="flex-1" />
      <Button
        onClick={handlePolishing}
        disabled={isLoading || !inputText?.trim()}
        className="relative ml-2 h-9 px-3 text-ms
          bg-slate-200 dark:bg-slate-700 hover:bg-[#f9fafb] dark:hover:bg-[#2e3238]"
      >
        <Wand2Icon className="h-4 w-4 text-[#8b5cf6] dark:text-[#a78bfa] opacity-80" />
        {isPolishingLoading ? <span>{IS_POLISHING}</span> : <span>{POLISH}</span>}
        {isPolishingLoading && (
          <Progress
            value={polishingProgress}
            className="absolute bottom-0 left-0 right-0 h-1 bg-[#e5e7eb] dark:bg-[#2e3238]"
            indicatorClassName="bg-[#8b5cf6] dark:bg-[#a78bfa]"
          />
        )}
      </Button>
      <Button
        onClick={handleTranslation}
        disabled={isLoading || !inputText?.trim()}
        className="relative ml-2 h-9 px-3 text-ms
          bg-slate-200 dark:bg-slate-700 hover:bg-[#f9fafb] dark:hover:bg-[#2e3238]"
      >
        <Languages className="h-4 w-4 text-[#8b5cf6] dark:text-[#a78bfa] opacity-80" />
        {isTranslationLoading ? <span>{IS_TRANSLATING}</span> : <span>{TRANSLATION}</span>}
        {isTranslationLoading && (
          <Progress
            value={translationProgress}
            className="absolute bottom-0 left-0 right-0 h-1 bg-[#e5e7eb] dark:bg-[#2e3238]"
            indicatorClassName="bg-[#8b5cf6] dark:bg-[#a78bfa]"
          />
        )}
      </Button>
      <Button
        onClick={handleReply}
        data-testid="reply-button"
        disabled={isLoading || !inputText?.trim()}
        className="relative overflow-hidden ml-2 h-9 px-3 text-ms text-white font-normal
          rounded-md shadow-sm hover:shadow
          bg-[#8b5cf6] hover:bg-[#7c3aed] dark:bg-[#a78bfa] dark:hover:bg-[#9061f9]"
      >
        {isReplyLoading ? <span>{IS_GENERATING}</span> : <span>{GENERATE}</span>}
        {isReplyLoading && (
          <Progress
            value={replyProgress}
            className="absolute bottom-0 left-0 right-0 h-1 bg-[#8b5cf6] dark:bg-[#a78bfa]"
            indicatorClassName="bg-[#e5e7eb] dark:bg-[#2e3238]"
          />
        )}
      </Button>
    </div>
  );
}
