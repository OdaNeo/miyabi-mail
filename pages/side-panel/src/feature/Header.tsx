import {
  Button,
  Tooltip,
  TooltipContent,
  TooltipProvider,
  TooltipTrigger,
  Popover,
  PopoverContent,
  PopoverTrigger,
} from '@extension/ui';
import { Settings, Sun, Moon, Globe, Feather, Clock } from 'lucide-react';
import { useStorage } from '@extension/shared';
import { darkModeStorage, i18nStorage } from '@extension/storage';
import { useOpenStore } from '@/store/openStore';
import { useI18n } from '@/hooks/useI18n';
import { SettingPopoverContent } from './SettingPopoverContent';
import { useHistoryOpenStore } from '@/store/historyOpenStore';

export function Header() {
  const darkMode = useStorage(darkModeStorage);

  const { toggleIsHistoryOpen } = useHistoryOpenStore();

  const { isOpen, setIsOpen } = useOpenStore();

  const { CHANGE_TO_LIGHT, CHANGE_TO_DARK, CHANGE_LANGUAGE } = useI18n();

  const handleSwitchLanguage = () => {
    i18nStorage.next();
  };

  const handleToggleDarkMode = () => {
    document.documentElement.classList.toggle('dark');
    darkModeStorage.set(!darkMode);
  };

  const handleTogglePopover = () => {
    setIsOpen(!isOpen);
  };

  return (
    <div className="flex justify-between items-center px-5 py-3 bg-gradient-to-r from-[#e9ecf1] to-[#f0f2f5] dark:from-[#23252b] dark:to-[#282a32] border-b border-[#e5e7eb] dark:border-[#2e3238]">
      <Feather className="w-5 h-5 text-[#8b5cf6] dark:text-[#a78bfa] opacity-80" />
      <h1 className="text-xl font-medium tracking-widest flex items-center ml-2">雅返信</h1>
      <div className="flex-1" />
      <div className="flex space-x-2">
        <Button
          variant="ghost"
          size="icon"
          className="w-8 h-8 text-slate-600 dark:text-slate-400 hover:bg-[#e2e3e4] dark:hover:bg-[#2e3238]"
          onClick={() => {
            toggleIsHistoryOpen();
          }}
        >
          <Clock data-testid="clock-icon" />
        </Button>
        <TooltipProvider>
          <Tooltip>
            <TooltipTrigger asChild>
              <Button
                variant="ghost"
                size="icon"
                className="w-8 h-8 text-slate-600 dark:text-slate-400 hover:bg-[#e2e3e4] dark:hover:bg-[#2e3238]"
                onClick={handleSwitchLanguage}
              >
                <Globe data-testid="globe-icon" />
              </Button>
            </TooltipTrigger>
            <TooltipContent className="bg-slate-700 text-slate-100 dark:bg-slate-200 dark:text-slate-800">
              {CHANGE_LANGUAGE}
            </TooltipContent>
          </Tooltip>
        </TooltipProvider>
        <TooltipProvider>
          <Tooltip>
            <TooltipTrigger asChild>
              <Button
                variant="ghost"
                size="icon"
                data-testid="toggle-dark-mode"
                className="w-8 h-8 text-slate-600 dark:text-slate-400 hover:bg-[#e2e3e4] dark:hover:bg-[#2e3238]"
                onClick={handleToggleDarkMode}
              >
                {darkMode ? <Sun className="h-4 w-4" /> : <Moon className="h-4 w-4" />}
              </Button>
            </TooltipTrigger>
            <TooltipContent className="bg-slate-700 text-slate-100 dark:bg-slate-200 dark:text-slate-800">
              <p>{darkMode ? CHANGE_TO_LIGHT : CHANGE_TO_DARK}</p>
            </TooltipContent>
          </Tooltip>
        </TooltipProvider>
        <Popover open={isOpen} onOpenChange={setIsOpen}>
          <PopoverTrigger asChild>
            <Button
              onClick={handleTogglePopover}
              variant="ghost"
              size="icon"
              data-testid="setting-icon"
              className="w-8 h-8 text-slate-600 dark:text-slate-400 hover:bg-[#e2e3e4] dark:hover:bg-[#2e3238]"
            >
              <Settings className="h-4 w-4" />
            </Button>
          </PopoverTrigger>
          <PopoverContent className="w-60 rounded-lg bg-white dark:bg-slate-900 border border-slate-200 dark:border-slate-700">
            <SettingPopoverContent setIsOpen={setIsOpen} />
          </PopoverContent>
        </Popover>
      </div>
    </div>
  );
}
