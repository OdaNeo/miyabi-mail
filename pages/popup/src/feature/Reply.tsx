import type { PROMPT_KEYS } from '@/utils/tts';
import { AnimatePresence, motion } from 'framer-motion';
import { Label, Button, cn } from '@extension/ui';
import { ChevronDown, ChevronUp, ClipboardCopy } from 'lucide-react';
import { useState } from 'react';
import { MessageBox } from '../components/MessageBox';
import { replyStorage, translationStorage } from '@extension/storage';
import { useStorage } from '@extension/shared';
import { useGeneratedStore } from '@/store/generatedStore';
import { useExpandedSectionStore } from '@/store/expandedSectionStore';
import { useI18n } from '@/hooks/useI18n';

export function Reply() {
  const reply = useStorage(replyStorage);
  const translation = useStorage(translationStorage);

  const { isGenerated } = useGeneratedStore();
  const { expandedSection, setExpandedSection } = useExpandedSectionStore();

  const [copied, setCopied] = useState(false);
  const { ORIGINAL_TRANSLATION, REPLY, COPY, COPIED } = useI18n();

  const toggleSection = (section: PROMPT_KEYS) => {
    setExpandedSection(expandedSection === section ? null : section);
  };

  const handleCopy = () => {
    navigator.clipboard.writeText(reply).then(() => {
      setCopied(true);
      setTimeout(() => setCopied(false), 2000);
    });
  };

  return (
    <AnimatePresence>
      {isGenerated && (
        <motion.div
          initial={{ opacity: 0, height: 0 }}
          animate={{ opacity: 1, height: 'auto' }}
          exit={{ opacity: 0, height: 0 }}
          className="space-y-2 overflow-y-auto scrollbar-custom"
        >
          {(['TRANSLATION', 'REPLY'] as PROMPT_KEYS[]).map(section => (
            <div key={section} className="space-y-1">
              <div
                className="flex justify-between items-center cursor-pointer h-8 p-1 rounded hover:bg-slate-100 dark:hover:bg-slate-800 transition-colors"
                onClick={() => toggleSection(section)}
                onKeyUp={e => {
                  if (e.key === 'Enter' || e.key === ' ') {
                    toggleSection(section);
                  }
                }}
                role="button"
                tabIndex={0}
              >
                <Label className="font-medium text-slate-700 dark:text-slate-300">
                  {section === 'TRANSLATION' ? ORIGINAL_TRANSLATION : REPLY}
                </Label>
                <div className="flex items-center space-x-1">
                  {section === 'REPLY' && (
                    <Button
                      variant="outline"
                      size="sm"
                      data-testid="copy-button"
                      onClick={e => {
                        e.stopPropagation();
                        handleCopy();
                      }}
                      className={cn(
                        'h-6 text-sm transition-colors',
                        copied
                          ? 'bg-emerald-500 text-white hover:bg-emerald-600'
                          : 'text-slate-600 dark:text-slate-400',
                      )}
                    >
                      {copied ? COPIED : COPY}
                      <ClipboardCopy className="h-3 w-3 ml-1" />
                    </Button>
                  )}
                  {expandedSection === section ? (
                    <ChevronUp className="h-3 w-3 text-slate-500 dark:text-slate-400" />
                  ) : (
                    <ChevronDown className="h-3 w-3 text-slate-500 dark:text-slate-400" />
                  )}
                </div>
              </div>
              {expandedSection === section && (
                <MessageBox>{section === 'TRANSLATION' ? translation : reply}</MessageBox>
              )}
            </div>
          ))}
        </motion.div>
      )}
    </AnimatePresence>
  );
}
