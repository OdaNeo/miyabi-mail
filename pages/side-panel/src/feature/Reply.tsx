import type { PROMPT_KEYS } from '@/utils/tts';
import { AnimatePresence, motion } from 'framer-motion';
import { Label } from '@extension/ui';
import { ChevronDown, ChevronUp } from 'lucide-react';
import { MessageBox } from '../components/MessageBox';
import { replyStorage, translationStorage } from '@extension/storage';
import { useStorage } from '@extension/shared';
import { useGeneratedStore } from '@/store/generatedStore';
import { useExpandedSectionStore } from '@/store/expandedSectionStore';
import { useI18n } from '@/hooks/useI18n';
import { CopyButton } from '@/components/CopyButton';

export function Reply() {
  const reply = useStorage(replyStorage);
  const translation = useStorage(translationStorage);

  const { isGenerated } = useGeneratedStore();
  const { expandedSection, setExpandedSection } = useExpandedSectionStore();

  const { ORIGINAL_TRANSLATION, REPLY } = useI18n();

  const toggleSection = (section: PROMPT_KEYS) => {
    setExpandedSection(expandedSection === section ? null : section);
  };

  const handleCopy = (section: PROMPT_KEYS) => {
    navigator.clipboard.writeText(section === 'REPLY' ? reply : translation);
  };

  return (
    <AnimatePresence>
      {isGenerated && (
        <motion.div
          initial={{ opacity: 0, height: 0 }}
          animate={{ opacity: 1, height: 'auto' }}
          exit={{ opacity: 0, height: 0 }}
          className="overflow-y-auto scrollbar-custom rounded-lg border border-[#e5e7eb] dark:border-[#2e3238]
        dark:bg-slate-800 dark:text-slate-100 dark:placeholder-slate-400 bg-white text-slate-900 placeholder-slate-500"
        >
          {(['TRANSLATION', 'REPLY'] as PROMPT_KEYS[]).map(section => (
            <div key={section}>
              <div
                className="flex justify-between items-center px-3 h-9 cursor-pointer bg-slate-100 dark:bg-[#23252b]
                 hover:bg-slate-200 dark:hover:bg-slate-800"
                onClick={() => toggleSection(section)}
                onKeyUp={e => {
                  if (e.key === 'Enter' || e.key === ' ') {
                    toggleSection(section);
                  }
                }}
                data-testid={`section-${section.toLocaleLowerCase()}`}
                role="button"
                tabIndex={0}
              >
                <Label className="font-medium text-[#4b5563] dark:text-[#d1d5db]">
                  {section === 'TRANSLATION' ? ORIGINAL_TRANSLATION : REPLY}
                </Label>
                <div className="flex items-center space-x-2">
                  <CopyButton handleCopyText={() => handleCopy(section)} />
                  {expandedSection === section ? (
                    <ChevronUp className="h-3 w-3 text-slate-500 dark:text-slate-400" />
                  ) : (
                    <ChevronDown className="h-3 w-3 text-slate-500 dark:text-slate-400" />
                  )}
                </div>
              </div>
              {expandedSection === section && (
                <MessageBox>{section === 'TRANSLATION' ? translation : reply}</MessageBox>
              )}
            </div>
          ))}
        </motion.div>
      )}
    </AnimatePresence>
  );
}
