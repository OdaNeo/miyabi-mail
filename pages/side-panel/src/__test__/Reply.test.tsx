import { render, fireEvent, screen } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { Reply } from '../feature/Reply';
import { useStorage } from '@extension/shared';
import type { PROMPT_KEYS } from '@/utils/tts';

vi.mock('framer-motion', () => ({
  motion: {
    div: ({ children, ...props }: any) => <div {...props}>{children}</div>,
  },
  AnimatePresence: ({ children }: any) => <>{children}</>,
}));

vi.mock('@extension/shared', () => ({
  useStorage: vi.fn(),
}));

const setExpandedSectionMock = vi.fn();

let mockExpandedSection: PROMPT_KEYS | null = null;

vi.mock('@/store/expandedSectionStore', () => ({
  useExpandedSectionStore: () => ({
    expandedSection: mockExpandedSection,
    setExpandedSection: setExpandedSectionMock,
  }),
}));

vi.mock('@/store/generatedStore', () => ({
  useGeneratedStore: () => ({
    isGenerated: true,
  }),
}));

vi.mock('@/hooks/useI18n', () => ({
  useI18n: () => ({
    ORIGINAL_TRANSLATION: 'Original Translation',
    REPLY: 'Reply',
    COPY: 'Copy',
    COPIED: 'Copied',
  }),
}));

const mockClipboard = {
  writeText: vi.fn(() => Promise.resolve()),
};

Object.assign(navigator, { clipboard: mockClipboard });

describe('Reply Component', () => {
  const useStorageMock = vi.mocked(useStorage);

  beforeEach(() => {
    vi.clearAllMocks();
    vi.useFakeTimers();
    mockExpandedSection = null;
  });

  afterEach(() => {
    vi.runOnlyPendingTimers();
    vi.useRealTimers();
  });

  it('should render correctly when generated', () => {
    useStorageMock
      .mockReturnValueOnce('Reply text') // reply
      .mockReturnValueOnce('Translation text'); // translation
    mockExpandedSection = 'TRANSLATION';
    const { container } = render(<Reply />);
    expect(container).toMatchSnapshot();
  });

  it('should toggle sections when clicking headers', () => {
    useStorageMock
      .mockReturnValueOnce('Reply text') // reply
      .mockReturnValueOnce('Translation text');
    render(<Reply />);
    fireEvent.click(screen.getByText('Original Translation'));
    expect(setExpandedSectionMock).toHaveBeenCalledWith('TRANSLATION');
    fireEvent.click(screen.getByText('Reply'));
    expect(setExpandedSectionMock).toHaveBeenCalledWith('REPLY');
  });

  it('should handle keyboard navigation', () => {
    useStorageMock
      .mockReturnValueOnce('Reply text') // reply
      .mockReturnValueOnce('Translation text');
    render(<Reply />);
    fireEvent.keyUp(screen.getByText('Original Translation'), { key: 'Enter' });
    expect(setExpandedSectionMock).toHaveBeenCalledWith('TRANSLATION');
    fireEvent.keyUp(screen.getByText('Reply'), { key: ' ' });
    expect(setExpandedSectionMock).toHaveBeenCalledWith('REPLY');
  });

  it('should not propagate click event when copying', () => {
    useStorageMock
      .mockReturnValueOnce('Reply text') // reply
      .mockReturnValueOnce('Translation text');
    render(<Reply />);
    const copyButton = screen.getAllByTestId('copy-icon-group');
    fireEvent.click(copyButton[1], { stopPropagation: true });
    expect(setExpandedSectionMock).not.toHaveBeenCalled();
    expect(mockClipboard.writeText).toHaveBeenCalled();
  });

  it('should copy text to clipboard', () => {
    useStorageMock
      .mockReturnValueOnce('Reply text') // reply
      .mockReturnValueOnce('Translation text');
    mockExpandedSection = 'REPLY';
    render(<Reply />);
    const copyButton = screen.getAllByTestId('copy-icon-group');
    fireEvent.click(copyButton[1]);
    expect(mockClipboard.writeText).toHaveBeenCalledWith('Reply text');
  });

  it('should copy when translation button is clicked', () => {
    useStorageMock
      .mockReturnValueOnce('Reply text') // reply
      .mockReturnValueOnce('Translation text');
    mockExpandedSection = 'TRANSLATION';
    render(<Reply />);
    const copyButton = screen.getAllByTestId('copy-icon-group');
    fireEvent.click(copyButton[0]);
    expect(mockClipboard.writeText).toHaveBeenCalledWith('Translation text');
  });

  it('should toggle sections when clicking icons', () => {
    useStorageMock
      .mockReturnValueOnce('Reply text') // reply
      .mockReturnValueOnce('Translation text');
    mockExpandedSection = 'REPLY';
    render(<Reply />);
    const toggleElement = screen.getByTestId('section-reply');
    fireEvent.click(toggleElement);
    expect(setExpandedSectionMock).toHaveBeenCalledWith(null);
  });
});
