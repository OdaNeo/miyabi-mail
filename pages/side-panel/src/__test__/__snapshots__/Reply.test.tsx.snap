// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Reply Component > should render correctly when generated 1`] = `
<div>
  <div
    animate="[object Object]"
    class="overflow-y-auto scrollbar-custom rounded-lg border border-[#e5e7eb] dark:border-[#2e3238] dark:bg-slate-800 dark:text-slate-100 dark:placeholder-slate-400 bg-white text-slate-900 placeholder-slate-500"
    exit="[object Object]"
    initial="[object Object]"
  >
    <div>
      <div
        class="flex justify-between items-center px-3 h-9 cursor-pointer bg-slate-100 dark:bg-[#23252b] hover:bg-slate-200 dark:hover:bg-slate-800"
        data-testid="section-translation"
        role="button"
        tabindex="0"
      >
        <label
          class="text-sm leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 font-medium text-[#4b5563] dark:text-[#d1d5db]"
        >
          Original Translation
        </label>
        <div
          class="flex items-center space-x-2"
        >
          <button
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent h-6 w-6 text-slate-500 hover:text-slate-900 relative"
            data-testid="copy-icon-group"
          >
            <div
              animate="[object Object]"
              class="absolute inset-0 flex items-center justify-center"
              exit="[object Object]"
              initial="[object Object]"
              transition="[object Object]"
            >
              <svg
                class="lucide lucide-copy h-4 w-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="14"
                  rx="2"
                  ry="2"
                  width="14"
                  x="8"
                  y="8"
                />
                <path
                  d="M4 16c-1.1 0-2-.9-2-2V4c0-1.1.9-2 2-2h10c1.1 0 2 .9 2 2"
                />
              </svg>
            </div>
          </button>
          <svg
            class="lucide lucide-chevron-up h-3 w-3 text-slate-500 dark:text-slate-400"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m18 15-6-6-6 6"
            />
          </svg>
        </div>
      </div>
      <div
        class="h-auto overflow-y-auto text-sm scrollbar-custom dark:bg-slate-800 dark:text-slate-100 dark:placeholder-slate-400 bg-white text-slate-900 placeholder-slate-500 p-3"
      >
        Translation text
      </div>
    </div>
    <div>
      <div
        class="flex justify-between items-center px-3 h-9 cursor-pointer bg-slate-100 dark:bg-[#23252b] hover:bg-slate-200 dark:hover:bg-slate-800"
        data-testid="section-reply"
        role="button"
        tabindex="0"
      >
        <label
          class="text-sm leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 font-medium text-[#4b5563] dark:text-[#d1d5db]"
        >
          Reply
        </label>
        <div
          class="flex items-center space-x-2"
        >
          <button
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent h-6 w-6 text-slate-500 hover:text-slate-900 relative"
            data-testid="copy-icon-group"
          >
            <div
              animate="[object Object]"
              class="absolute inset-0 flex items-center justify-center"
              exit="[object Object]"
              initial="[object Object]"
              transition="[object Object]"
            >
              <svg
                class="lucide lucide-copy h-4 w-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="14"
                  rx="2"
                  ry="2"
                  width="14"
                  x="8"
                  y="8"
                />
                <path
                  d="M4 16c-1.1 0-2-.9-2-2V4c0-1.1.9-2 2-2h10c1.1 0 2 .9 2 2"
                />
              </svg>
            </div>
          </button>
          <svg
            class="lucide lucide-chevron-down h-3 w-3 text-slate-500 dark:text-slate-400"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
